models/agent.py

```diff
@@ -0,0 +1,189 @@
+import numpy as np
+import torch
+from torch.distributions import Normal
+
+
+class RLAgent:
+    def __init__(self, policy_type, **policy_kwargs):
+        self.policy_type = policy_type
+        self.policy = self._get_policy(policy_type, **policy_kwargs)
+        self.on_policy = policy_type in ['ppo']
+
+    def to(self, device):
+        self.device = device
+        self.policy.to(device)
+        return self
+
+    def init_optimizers(self, **lrs):
+        self.policy.init_optimizers(**lrs)
+
+    def update(self, buffer, batch_size, num_updates, **kwargs):
+        return self.policy.update(buffer, batch_size, num_updates, **kwargs)
+
+    def get_action(self, state, deterministic=True):
+        state = torch.as_tensor(state, dtype=torch.float32, device=self.device)
+        mean, logstd = self.policy(state.unsqueeze(0))
+        mean, logstd = mean.squeeze(0), logstd.squeeze(0)
+
+        if deterministic:
+            action = mean
+            logprob = torch.zeros_like(action, device=self.device).sum(-1)
+        else:
+            dist = Normal(mean, logstd.exp())
+            action = dist.rsample()
+            logprob = dist.log_prob(action).sum(-1)
+
+        return action, logprob
+
+    @torch.no_grad()
+    def rollout_step(self, env, episode_step, observation, rand=False, deterministic=False):
+        if rand:
+            action = env.action_space.sample()
+            logprob = np.zeros_like(action).sum(-1)
+        else:
+            action, logprob = self.get_action(observation, deterministic=deterministic)
+            action, logprob = action.cpu().numpy(), logprob.cpu().numpy()
+        action = np.clip(action, env.action_space.low, env.action_space.high)
```

Comment:
Q) Is clipping same as clamping? I've seen this expression(clipping) in several posts, and I feel like it is identical to clamping.

Reply:
`clip` is for numpy, and `clamp` is for torch.